### ISSUE TITLE
[Editor] Camera rotation broken

### ISSUE THREAD

**Issue by SuperWangKai** [urho3d/Urho3D#2039]
In the current implementation of editor view, if we rotate the camera, it will focus to the selection, and rotate around the center of it. This can easily loose current point of interest and troublesome when we are editing large scale of scene. We need to navigate back to the previous point of interest after every rotation.

For most of cases, we just want to rotate the camera and not want to "Rotate Around Selection"(it's an option of Blender).

Thanks!

**Comment by 1vanK** [comment]
use RMB for rotation

**Comment by SuperWangKai** [comment]
@1vanK 
RMB rotation is another way of rotation. It allows the camera rotate around it's own center.

**Comment by 1vanK** [comment]
strafe keys (A, W) + mouse allows you rotate around  any target like shooter game :)

**Comment by SuperWangKai** [comment]
@1vanK Yes, you are right, however, I think if we still need the rotation more natural just like Blender or Unity Editor has.

**Comment by SuperWangKai** [comment]
I realized that someone should do something. So I spent many weekends to fix this issue. I just could not stop to improve the camera as a whole. So many other improvements also added.
Here is my PR  - https://github.com/urho3d/Urho3D/pull/2115

**Comment by 1vanK** [comment]
I haven't watched PR yet. Does this replace the existing way that everyone is used to? Or it is additional option?

**Comment by SuperWangKai** [comment]
For the camera movements, you can still use FPS mode for both standard and Blende modes. By adding a look-at point, camera just feel more natural and focus will no longer lost. Camera looking for objects is better, more accurate and smooth. Top/Font/Side view switching also improved.

**Comment by 1vanK** [comment]
I'm completely satisfied with the rotation around the selected object. I have been using a blender for a long time and I always hated rotating around a random point of space and the need to constantly press key to focus on selected object

**Comment by SuperWangKai** [comment]
I don't want to start a flight here. There are many reasons why this is important. 3ds max, blender, Unity, Unreal Engine 4, named just a little. using this way and ppl get used to it.

**Comment by 1vanK** [comment]
Just add it as option

**Comment by SuperWangKai** [comment]
The "Rotate Around Selection" of Blender is actually very different from the current Urho Editor's. Blender won't change the look-at position (view center) when we orbiting the camera. However, for Urho, when we orbiting the camera, it just jumps to the selected even if the selection is miles away from what we are currently looking at!

Urho's current implementation is just too strange to me and also to most ppl I am afraid. With the improved look-close function (F/Home for Standard Hotkey, Home/NumPad Period for Blender Hotkey, or by double clicking in the Hierarchy Window) you will be totally fine. Actually, even if you moved the camera a little bit after look-close, you won't lost your focus of current selection. I don't see why we need to rotate strictly around the center of the selected.

For the Top/Front/Side view change (NumPad 1, 3, 7 , 9 (9 was missing)), if we don't have a look-at point, it also very very strange. You can check the improvement.

**Comment by 1vanK** [comment]
Understand me, please. I don't mind this change. To the editor was added a lot of things that seem more convenient to someone  Ffor example "New node mode" from me and "Blender mode" from MonkieyFirst), but they are off by default.. Hundreds of people use the editor and have skills. And then a new person comes in and says - your skills are wrong.

**Comment by SuperWangKai** [comment]
IMO, there are a lot work that can be and should be done for the editor. Unless we are ready for reinventing the wheels and building a better new one.

Currently the editor is even not ready for production if we use industry standard. The camera jumping issue should be treated as a bug instead of feature.

**Comment by eugeneko** [comment]
> Currently the editor is even not ready for production if we use industry standard

Editor was never _intended_ to be production-ready. It is closer to scripting sample than Editor like Unity/etc.
There were several attempts to do something with the Editor and I am going to do another attempt in the future.

**Comment by 1vanK** [comment]
Anyway Editor often required modification for specific projects. It's done very quickly and conveniently. For example I add buttons to add collections of objects to scene with random rotation. So do not hesitate to have your own version of the editor such that you like.

**Comment by weitjong** [comment]
First of all, thanks for your time for this PR. At this point adding the fix/feature as a new option sounds like a good compromise. No matter how bad it is with the current implementation, some people already get used to it.

Our community is small as it is. We need to pull all the good bits together. We should encourage people to submit more PR to share their work and not the other way. So, thanks again for sharing your work.

**Comment by SuperWangKai** [comment]
Thank you @weitjong for your positive reply. 

Some people's responses are just too negative. Too many argues here and there around Urho these days. I though I could contribute somehow, if PR is rejected I'll just be happy to keep it for my own use. 

Please try it before you say no to something.

Just download the bin.zip, extract it and override your original bin folder - [bin.zip](https://github.com/SuperWangKai/UrhoTestStorage/blob/master/bin.zip)

Now, I really don't understand why people are so negative and preventing others to improve the editor. I'm afraid I have to say something to defend for the right thing.

I insist that the look-at point missing is a bug. Historically, the reason why Urho Editor use 'Jump to Selection' when rotating or use the selection center for Top/Side/Front view switching is a mistake I guess. Since you eventually have to find a pivot for rotation, for the current editor, we have only two choices: camera center or selection center. But now, we have a look-at point which is very consistent the camera can behave just like other normal editors (3ds max, maya, blender, Unity, UE4...). I believe that there are many people from other communities like Unity will feel as frustrated just like me.

Adding a look-at point, basically means every pan/orbiting/zoom code piece has to be modified. I don't think make my modification as an option is a good idea. It's beyond my ability to squeeze a new mechanism here and there into the current code which cost me days to refactor and IMHO that work is not necessary.

**However, I can add an option for the legacy by simulating current camera jumping behavior.** Maybe we could call it "Jump to selection for camera rotation" or something sounds better. But to be honest, I really doubt that how many users will miss the current broken implementation (Maybe I should move to the forum and start a poll).

Before any new working Editor shows, current Editor is still the best choice. I saw people start a new editor project and throw it away. The problem here is that it takes a lot of time to develop a working, productive Editor. This community has invested many many hours into the current editor and we should keep making it better. As I remembered, it took me a full year to develop a serious and fully functional editor for a commercial studio years ago.

**Comment by 1vanK** [comment]
When you move camera, z point of "center of rotation" is just random point. When I work with object, I rotate around this object. Why you need to rotate around an empty point of space? You do not compromise and do not take into account the opinions of others. Instead of adding an option you want to break something that others use and make them do as you see fit. After it you accuse someone for negativity. It's epic.

**Comment by SuperWangKai** [comment]
> When you move camera, z point of "center of rotation" is just random value. When I work with object, I rotate around this object. Why you need to rotate around an empty point of space? 

There are many reasons why we need a look-at point. Please ask 3ds max, maya, blender, Unity, UE4... the same question. I think I have explained a lot in this thread.

> Instead of adding an option you want to break something that others use and make them do as you see fit

What I'm saying is that I can make the fix as the default and make the current way (I think it's a bug) as an option.
Nothing will be changed except making the "camera jumping" as an option (this is my compromise). You have same FPS moving, pan, RMB orbit, just as before.

I'm sorry but I didn't see any other ppl against my PR, and I'll just stop arguing this topic with you  @1vanK . And I won't bother myself to PR (advertise) my PR in this community which is really pathetic.
 Good is good, bad is bad. Let other developers/users decide.

**Comment by 1vanK** [comment]
> Nothing will be changed except making the "camera jumping" as an option (this is my compromise). You have same FPS moving, pan, RMB orbit, just as before.

Thank you for allowing me, thanks

> I'm sorry but I didn't see any other ppl against my PR

I saw it, but in a very polite way. We have a very polite community (except me, because I'm a crazy Russian, do not pay attention to me and send more PR)

**Comment by SuperWangKai** [comment]
All done. You can check my final pull Request - https://github.com/urho3d/Urho3D/pull/2115

**Comment by andrekoehler** [comment]
Just tried out the new editor version and enabled "Rotate Around Selection" to use the old middle mouse button rotation behavior. I noticed one change that should be reverted IMO: Before the camera changes were merged, the editor would first calculate the current distance of the camera from the selected object and then keep this distance while orbiting around the object. Now after the patch the distance in which the camera rotates around the selected object is always reset to some constant value. This is very annoying when I want to inspect an object from close up. I would strongly suggest to change the camera patch to re-add this old behavior when "Rotate Around Selection" is enabled.
